flocker/node/agents/test/test_blockdevice.py

```diff
@@ -5249,3 +5257,231 @@ def test_get_device_path_until_detach(self):
+            return d
+
+        def loop_until_converged():
+            result = converge()
+
+            def repeat_if_necessary(changes_needed):
+                if changes_needed:
+                    return loop_until_converged()
+                else:
+                    return None
+
+            result.addCallback(repeat_if_necessary)
+            return result
+        return loop_until_converged()
+
+    def test_protect_against_docker(self):
+        """
+        Verify that in a standard lifecycle of a dataset there is never a time
+        after the path of a dataset is reported to the control node that a
+        successful write to that path is not persisted to the backing loop
+        device.
+        """
+        evaluate_function = None
+
+        def callback(when_description):
+            if evaluate_function:
+                evaluate_function(when_description)
+        actual_blockdevice_manager = BlockDeviceManager()
+        proxy_blockdevice_manager = CallbackProxy(
+            actual_blockdevice_manager, callback)
+
+        # Set up a deployer with a blockdevice_manager that has hooks so we can
+        # inject code around interactions with the blockdevices on the system.
+        deployer = create_blockdevicedeployer(
+            self, node_uuid=self.NODE_UUID,
+            block_device_manager=proxy_blockdevice_manager)
+
+        testing_mountroot = mountroot_for_test(self)
+
+        def has_file_named(blockdevice, filename):
+            """
+            Returns whether a given blockdevice has a file with the given name.
+            """
+            test_mountpoint = testing_mountroot.child(str(uuid4()))
+            test_mountpoint.makedirs()
+            actual_blockdevice_manager.mount(blockdevice, test_mountpoint)
+            result = test_mountpoint.child(filename).exists()
+            actual_blockdevice_manager.unmount(blockdevice)
+            test_mountpoint.remove()
+            return result
+
+        # This is information about the dataset that is intentionally only
+        # populated once the path of the dataset is known by the cluster and
+        # thus could be known by an API client.
+        dataset_info_from_cluster = {
+            # The path of the dataset.
+            'paths': set(),
+            # The blockdevice of the dataset. This is used for back door
+            # verification that writes are persisted to the blockdevice.
+            'dev': None,
+        }
+
+        def invariant(when_description):
+            """
+            An invariant that should always hold. Specifically, after the path
+            to a dataset is known by the cluster then writes to that path
+            should either succeed and be persisted to the underlying block
+            device, or the write should fail.
+
+            Returns False if none of the known paths of the dataset can be
+            written to.
+            """
+            write_success = []
```

Comment:
So it's a list in case we ever have 2 different paths for a dataset at 2 different times. Current implementation does not do that, but if you imagine an implementation where you mount datasets at random paths every time then for a single dataset after a manifest->non-manifest->manifest transition you could imagine there would be two paths that an external agent might try to write to for the dataset.